src/main/java/jenkins/scm/api/SCMFileSystem.java

```diff
@@ -388,6 +413,18 @@ public static boolean supports(@NonNull SCMSource source) {
          */
         public abstract boolean supports(SCMSource source);
 
+        /**
+         * Checks if this {@link Builder} supports the supplied {@link SCMSourceDescriptor}.
+         *
+         * @param descriptor the {@link SCMSourceDescriptor}
+         * @return {@code true} if and only if the supplied {@link SCMSourceDescriptor}'s {@link SCMSource} class is
+         * supported by this {@link Builder}, {@code false} if {@link #build(SCMSource, SCMHead, SCMRevision)} will
+         * <strong>always</strong> return {@code null}, and {@code false} by default for compatibility.
+         */
+        public boolean supports(SCMSourceDescriptor descriptor) {
+            return false;
```

Comment:
I think this might not be the correct value. Ideally what we would do is make this method `final` and make the `SCMSource` version non-abstract. Then we introduce `protected abstract` methods that the `public final` methods call (guarded by a reflective check of non-implemented)

This would allow us to try and infer if the answer *should* be `true` or `false` for a non-implemented method based on things like:
* Is it in the same package to the `SCMSource`
* Is the `SCMSource` an outer-class of this `Builder`

I think if either of those are the case and we are dealing with a legacy `Builder` implementation then we should return `true` and not `false`, otherwise we will have false negatives and you basically have to instantiate the `SCMSource` anyway

Reply:
That's reasonable - I did probably get the default value wrong. I'll think on the rest.

Reply:
To be clear, we only check heuristics if the `protected abstract boolean checkSupports(SCMSourceDescriptor);` has not been implemented. Which would be the case only for plugins compiled against the older versions of scm-api. As soon as you bump the scm-api dependency, you'd be forced to implement it and thus can return the correct `true`/`false` value and the heuristics can safely be ignored

Reply:
The heuristics will also need to handle inheritance - i.e., a specific `AbstractGitSCMSource` implementation may not have its own `SCMFileSystem.Builder` implementation, but it can still use `GitSCMFileSystem`.

Reply:
Wait, maybe not - since we're checking all `SCMFileSystem.Builder` implementations, we just need the `GitSCMFileSystem.Builder#supportsDescriptor` implementation to be implemented in that scenario.

Reply:
(all told, this just really makes me wish we had the `SCMSource` as a type parameter on `SCMFileSystem.Builder` or something like that. Ah well.)